pkg/flow/streaming/flow_sliding_window.go

```diff
@@ -0,0 +1,342 @@
+	// guard timerHeap
+	timerMu  sync.Mutex
+	trigger  EventTimeTrigger
+	aggrFunc flow.AggregateFunction
+
+	// For api.Operator
+	in           chan interface{}
+	out          chan interface{}
+	done         chan struct{}
+	purgedWindow chan timeWindow
+}
+
+func (s *SlidingTimeWindows) GetCurrentWatermark() int64 {
+	return s.currentWatermark
+}
+
+func (s *SlidingTimeWindows) RegisterEventTimeTimer(triggerTime int64) {
+	s.timerMu.Lock()
+	defer s.timerMu.Unlock()
+	heap.Push(s.timerHeap, &internalTimer{
+		triggerTimeMillis: triggerTime,
+		w:                 s.currentWindow,
+	})
+}
+
+func (s *SlidingTimeWindows) In() chan<- interface{} {
+	return s.in
+}
+
+func (s *SlidingTimeWindows) Out() <-chan interface{} {
+	return s.out
+}
+
+func (s *SlidingTimeWindows) Setup(ctx context.Context) error {
+	// start processing
+	go s.receive()
+	// start emitting
+	go s.emit()
+
+	return nil
+}
+
+func (s *SlidingTimeWindows) emit() {
+	s.Add(1)
+	defer s.Done()
+	for w := range s.purgedWindow {
+		s.queueMu.Lock()
+		// build a window slice and send it to the out chan
+		var windowBottomIndex int
+		windowUpperIndex := s.queue.Len()
+		slideUpperIndex := windowUpperIndex
+		slideUpperTime := w.start + s.slide
+		windowBottomTime := w.start
+		for i, item := range s.queue.Items {
+			if item.(*TimestampedValue).TimestampMillis() < windowBottomTime {
+				windowBottomIndex = i
+			}
+			if item.(*TimestampedValue).TimestampMillis() > slideUpperTime {
+				slideUpperIndex = i
+				break
+			}
+		}
+		slidingSlices := s.queue.Slice(windowBottomIndex, slideUpperIndex)
+		// if we've collected some items, reallocate queue
+		if len(slidingSlices) > 0 {
+			remainingItems := s.queue.Slice(slideUpperIndex, windowUpperIndex)
+			// reset the queue
+			var err error
+			s.queue, err = s.queue.WithNewItems(remainingItems)
+			if err != nil {
+				// TODO: drain error
+				panic("drain error")
+			}
+			heap.Init(s.queue)
+		}
+		s.queueMu.Unlock()
+
+		if len(slidingSlices) > 0 {
+			data := make([]interface{}, 0, len(slidingSlices))
+			for _, elem := range slidingSlices {
+				data = append(data, elem.(*TimestampedValue).Data())
+			}
+			s.out <- flow.NewStreamRecord(s.aggrFunc(data), slideUpperTime)
+		}
+	}
+}
+
+func (s *SlidingTimeWindows) purgeOutdatedWindows() {
+	s.timerMu.Lock()
+	defer s.timerMu.Unlock()
+	for {
+		if lookAhead, ok := s.timerHeap.Peek().(*internalTimer); ok {
+			if lookAhead.triggerTimeMillis <= s.currentWatermark {
+				oldestTimer := heap.Pop(s.timerHeap).(*internalTimer)
+				s.purgedWindow <- oldestTimer.w
+				continue
+			}
+		}
+		return
+	}
+}
+
+func (s *SlidingTimeWindows) receive() {
+	s.Add(1)
+	defer s.Done()
+	for elem := range s.in {
+		record := elem.(flow.StreamRecord)
+		// assume records are consumed one by one in strict time order
+		s.currentWatermark = record.TimestampMillis()
+		s.purgeOutdatedWindows()
+
+		assignedWindows, err := s.AssignWindows(record.TimestampMillis())
```

Comment:
It's tricky here. Why does it purge a window repeatly? The very beginning window in a interval gets more than one purging.